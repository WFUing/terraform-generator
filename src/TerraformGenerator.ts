import shell from 'shelljs';
import child_process from 'child_process';
import fs from 'fs';
import path from 'path';
import { Block, Resource, Data, Module, Output, Provider, Variable, Backend, ResourceToDataOptions } from '.';
import TerraformGeneratorUtils from './TerraformGeneratorUtils';

export type TerraformVersion = '0.11' | '0.12';

/**
 * @param version Terraform version
 */
export interface TerraformGeneratorOptions {
  version: TerraformVersion;
}

/**
 * @param dir directoty, default = .
 * @param filename Terraform filename, must ends with .tf, default = terraform.tf
 * @param format use 'terraform fmt' to format the configuration, Terraform must be installed, default = false
 */
export interface WriteOptions {
  dir?: string;
  filename?: string;
  format?: boolean;
}

export default class TerraformGenerator {

  private readonly options: TerraformGeneratorOptions;
  private readonly arguments: Record<string, any>;
  private readonly blocks: Block[] = [];

  /**
   * Construct Terraform generator.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param options options
   * @param args arguments
   */
  constructor(options: TerraformGeneratorOptions, args?: Record<string, any>) {
    this.options = options;
    this.arguments = args;
  }

  /**
   * Generate Terraform configuration as string.
   */
  generate(): string {
    let str = '';

    if (this.arguments || this.blocks.filter(block => block instanceof Backend).length > 0) {
      str += 'terraform {\n';
      str += TerraformGeneratorUtils.argumentsToString(this.options.version, this.arguments);
      this.blocks.forEach(block => {
        if (block instanceof Backend) {
          str += block.toTerraform(this.options.version);
        }
      });
      str += '}\n\n';
    }

    this.blocks.forEach(block => {
      if (!(block instanceof Backend)) {
        str += block.toTerraform(this.options.version);
      }
    });

    return str;
  }

  /**
   * Write Terraform configuration to a file.
   * 
   * @param options options
   */
  write(options?: WriteOptions): void {
    if (!options) {
      options = {};
    }
    if (!options.dir) {
      options.dir = '.';
    }
    if (!options.filename) {
      options.filename = 'terraform.tf';
    }
    if (!options.filename.endsWith('.tf')) {
      options.filename += '.tf';
    }

    shell.mkdir('-p', options.dir);
    fs.writeFileSync(path.join(options.dir, options.filename), this.generate());

    if (options.format) {
      child_process.execSync('terraform fmt', { cwd: options.dir });
    }
  }

  /**
   * Add blocks into Terraform.
   * 
   * @param blocks blocks
   */
  addBlocks(...blocks: Block[]): TerraformGenerator {
    blocks.forEach(block => this.blocks.push(block));
    return this;
  }

  /**
   * Add provider into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param args arguments
   */
  provider(type: string, args?: Record<string, any>): Provider {
    const block = new Provider(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add resource into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param name name
   * @param args arguments
   */
  resource(type: string, name: string, args?: Record<string, any>): Resource {
    const block = new Resource(type, name, args);
    this.addBlocks(block);
    return block;
  }
  /**
   * Convert resource into data source and add it into Terraform.
   * 
   * @param resource resource
   * @param options options
   * @param argNames names of resource arguments to be converted into data source arguments;
   * use array for name mapping, position 0 = original resource's argument name, position 1 = mapped data source's argument name
   * @param args extra arguments
   */
  dataFromResource(resource: Resource, options: ResourceToDataOptions, argNames: (string | [string, string])[], args?: Record<string, any>): Data {
    const block = resource.toData(options, argNames, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add data source into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param name name
   * @param args arguments
   */
  data(type: string, name: string, args?: Record<string, any>): Data {
    const block = new Data(type, name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add module into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  module(name: string, args?: Record<string, any>): Module {
    const block = new Module(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add output into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  output(name: string, args?: Record<string, any>): Output {
    const block = new Output(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add provider into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  variable(name: string, args?: Record<string, any>): Variable {
    const block = new Variable(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add backend into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param args arguments
   */
  backend(type: string, args?: Record<string, any>): Backend {
    const block = new Backend(type, args);
    this.addBlocks(block);
    return block;
  }

}
