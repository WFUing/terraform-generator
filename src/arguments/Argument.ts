export default class Argument {

  readonly argument: string | Argument;
  readonly asIs: boolean;

  /**
   * Construct argument.
   * 
   * @param arg argument as string or copy from another argument object
   * @param asIs argument will be printed as is, without extra symbol, quotes and whatnot, regardless of Terraform version, default = false.
   */
  constructor(arg: string | Argument, asIs = false) {
    if (!arg || (typeof arg === 'string' && !arg.trim())) {
      throw new Error('Argument cannot be empty.');
    }

    this.argument = typeof arg === 'string' ? arg.trim() : arg;
    this.asIs = asIs;
  }

  /**
   * To Terraform representation.
   */
  toTerraform(): string {
    let str = '';
    if (this.argument instanceof Argument) {
      str += this.argument.toTerraform();
    } else {
      str += this.argument;
    }
    return str;
  }

  /**
   * To string, used when argument is used in string interpolation.
   */
  toString(): string {
    return `\${${this.toTerraform()}}`;
  }

}

/**
 * Convenient function to construct new argument.
 * 
 * @param arg argument as string or copy from another argument object
 * @param asIs argument will be printed as is, without extra symbol, quotes and whatnot, regardless of Terraform version, default = false.
 */
export const arg = (arg: string | Argument, asIs = false): Argument => new Argument(arg, asIs);
