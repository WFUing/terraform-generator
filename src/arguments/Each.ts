import { Argument } from '..';

export default class Each extends Argument {

  /**
   * Construct each argument.
   * 
   * @param attrName attribute name
   */
  constructor(attrName: string) {
    super(`each.${attrName}`);
  }

}

/**
 * Convenient function to construct new each argument.
 * 
 * @param attrName attribute name
 */
export const each = (attrName: string): Each => new Each(attrName);
