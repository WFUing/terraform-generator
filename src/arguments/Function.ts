import { Argument } from '..';
import TerraformGeneratorUtils from '../TerraformGeneratorUtils';

export default class Fn extends Argument {

  /**
   * Construct function argument.
   * 
   * @param name function name
   * @param args function arguments
   */
  constructor(name: string, ...args: any[]) {
    super(Fn.constructArgument(name, ...args));
  }

  private static constructArgument(fn: string, ...args: any[]): string {
    if (!fn || !fn.trim()) {
      throw new Error('Function name cannot be empty.');
    }
    if (args.filter(arg => arg == null).length > 0) {
      throw new Error(`Invalid function argument: ${args}`);
    }

    let str = `${fn}(`;
    args.forEach((arg, i) => {
      str += TerraformGeneratorUtils.argumentValueToString(null, arg);
      if (i < args.length - 1) {
        str += ', ';
      }
    });
    str += ')';
    return str;
  }

}

/**
 * Convenient function to construct new function.
 * 
 * @param name function name
 * @param args function arguments
 */
export const fn = (name: string, ...args: any[]): Fn => new Fn(name, ...args);
