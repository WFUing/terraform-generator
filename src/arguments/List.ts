import { Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class List extends Argument {

  /**
   * Construct list.
   *
   * @param elements list elements
   */
  constructor(...elements: any[]) {
    super(List.#constructArgument(elements));
  }

  static #constructArgument(elements: any[]): string {
    let str = '[\n';
    elements.forEach((element, i) => {
      str += `${Util.argumentValueToString(element)}${i < elements.length - 1 ? ',' : ''}\n`;
    });
    str += ']';
    return str;
  }

}

/**
 * Convenient function to construct new [[List]].
 *
 * @param elements list elements
 *
 * @category Argument
 */
export const list = (...elements: any[]): List => new List(...elements);
