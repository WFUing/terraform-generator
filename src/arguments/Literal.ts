import { Argument } from '..';

export default class Literal extends Argument {

  /**
   * Construct literal argument.
   * 
   * @param literal literal string
   */
  constructor(literal: string) {
    super(literal, true);
  }

}

/**
 * Convenient function to construct new literal argument.
 * 
 * @param literal literal string
 */
export const literal = (literal: string): Literal => new Literal(literal);
