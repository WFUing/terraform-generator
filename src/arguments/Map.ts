import { TerraformArgs, Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Map extends Argument {

  readonly arguments: TerraformArgs;

  /**
   * Construct map.
   *
   * @param args map values
   */
  constructor(args: TerraformArgs) {
    super(Map.#constructArgument(args));
    this.arguments = args;
  }

  static #constructArgument(args: TerraformArgs): string {
    return Util.argumentValueToString(args)!;
  }

}

/**
 * Convenient function to construct new [[Map]].
 *
 * @param args map values
 *
 * @category Argument
 */
export const map = (args: TerraformArgs): Map => new Map(args);
