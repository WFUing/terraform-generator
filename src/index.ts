export { BlockArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, ProvisionerArgs, ProvisionerType,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc
} from './arguments';

export {
  List, list,
  Map, map
} from './types';
