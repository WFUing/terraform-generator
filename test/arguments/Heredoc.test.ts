import { Heredoc, heredoc } from '../../src/arguments';

test('Heredoc', () => {
  expect(new Heredoc('x').toTerraform()).toMatchSnapshot();
  expect(new Heredoc({ x: 123 }).toTerraform()).toMatchSnapshot();
});

test('heredoc', () => {
  expect(heredoc('x').toTerraform()).toMatchSnapshot();
  expect(heredoc({ x: 123 }).toTerraform()).toMatchSnapshot();
});
