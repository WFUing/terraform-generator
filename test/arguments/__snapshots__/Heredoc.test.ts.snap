// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heredoc 1`] = `
"<<EOT
x
EOT
"
`;

exports[`Heredoc 2`] = `
"<<EOT
{
  &tfgquot;x&tfgquot;: 123
}
EOT
"
`;

exports[`heredoc 1`] = `
"<<EOT
x
EOT
"
`;

exports[`heredoc 2`] = `
"<<EOT
{
  &tfgquot;x&tfgquot;: 123
}
EOT
"
`;
